From c76b2bf0b4e752455e5ffbd7dee12e935c2eaccd Mon Sep 17 00:00:00 2001
From: Andriy Gelman <andriy.gelman@gmail.com>
Date: Sat, 9 May 2020 19:34:06 -0400
Subject: [PATCH] avcodec/v4l2_context: Finish draining if V4L2_BUF_FLAG_LAST
 is set

V4L2 api can indicate that flushing of the capture buffers is completed
by setting the V4L2_BUF_FLAG_LAST flag.
Use guards because the flag was only defined in Linux v4.2.

Reference:
linux/Documentation/media/uapi/v4l/dev-decoder.rst

    "The client must continue to handle both queues independently,
     similarly to normal decode operation. This includes:
     ...
     - queuing and dequeuing CAPTURE buffers, until a buffer marked with
       the V4L2_BUF_FLAG_LAST flag is dequeued"

Reviewed-by: Ming Qian <ming.qian@nxp.com>
Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>
---
 libavcodec/v4l2_context.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git mythtv-old/mythtv/external/FFmpeg/libavcodec/v4l2_context.c mythtv-new/mythtv/external/FFmpeg/libavcodec/v4l2_context.c
index 52dedb119f3..29b144ed73e 100644
--- mythtv-old/mythtv/external/FFmpeg/libavcodec/v4l2_context.c
+++ mythtv-new/mythtv/external/FFmpeg/libavcodec/v4l2_context.c
@@ -412,6 +412,10 @@ static V4L2Buffer* v4l2_dequeue_v4l2buf(V4L2Context *ctx, int timeout)
                 ctx->done = 1;
                 return NULL;
             }
+#ifdef V4L2_BUF_FLAG_LAST
+            if (buf.flags & V4L2_BUF_FLAG_LAST)
+                ctx->done = 1;
+#endif
         }
 
         avbuf = &ctx->buffers[buf.index];
